src/simulation/dynamics/constraintEffector/_UnitTest/test_ConstraintDynamicEffectorUnit.py

```diff
@@ -0,0 +1,331 @@
+    combinedRotAngMom = rotAngMom1PntCT_N + rotAngMom2PntCT_N
+    combinedRotEnergy = rotEnergy1 + rotEnergy2
+
+    # Plotting
+    plt.close("all")
+    plt.figure()
+    plt.clf()
+    for i in range(3):
+        plt.semilogy(constraintTimeData, np.abs(psi_B1[:, i]))
+    plt.semilogy(constraintTimeData, np.linalg.norm(psi_B1,axis=1))
+    plt.legend([r'$\psi_1$',r'$\psi_2$',r'$\psi_3$',r'$\psi$ magnitude'])
+    plt.xlabel('time (seconds)')
+    plt.ylabel(r'variation from fixed position: $\psi$ (meters)')
+    plt.title('Direction Constraint Violation Components')
+
+    plt.figure()
+    plt.clf()
+    for i in range(3):
+        plt.semilogy(constraintTimeData, np.abs(4*np.arctan(sigma_B2B1[:, i]) * macros.R2D))
+    plt.semilogy(constraintTimeData, np.linalg.norm(4*np.arctan(sigma_B2B1) * macros.R2D,axis=1))
+    plt.legend([r'$\phi_1$',r'$\phi_2$',r'$\phi_3$',r'$\phi$ magnitude'])
+    plt.xlabel('time (seconds)')
+    plt.ylabel(r'relative attitude angle: $\phi$ (deg)')
+    plt.title('Attitude Constraint Violation Components')
+
+    plt.figure()
+    plt.clf()
+    plt.plot(conservationTimeData, (combinedRotAngMom[:,0] - combinedRotAngMom[0,0])/combinedRotAngMom[0,0],
+             conservationTimeData, (combinedRotAngMom[:,1] - combinedRotAngMom[0,1])/combinedRotAngMom[0,1],
+             conservationTimeData, (combinedRotAngMom[:,2] - combinedRotAngMom[0,2])/combinedRotAngMom[0,2])
+    plt.xlabel('time (seconds)')
+    plt.ylabel('Relative Difference')
+    plt.title('Combined Rotational Angular Momentum')
+
+    plt.figure()
+    plt.clf()
+    plt.plot(conservationTimeData, (combinedRotEnergy - combinedRotEnergy[0])/combinedRotEnergy[0])
+    plt.xlabel('time (seconds)')
+    plt.ylabel('Relative Difference')
+    plt.title('Combined Rotational Energy')
+
+    if testCase == 'Gravity':
+        plt.figure()
+        plt.clf()
+        plt.plot(conservationTimeData, (combinedOrbAngMom[:,0] - combinedOrbAngMom[0,0])/combinedOrbAngMom[0,0], 
+             conservationTimeData, (combinedOrbAngMom[:,1] - combinedOrbAngMom[0,1])/combinedOrbAngMom[0,1], 
+             conservationTimeData, (combinedOrbAngMom[:,2] - combinedOrbAngMom[0,2])/combinedOrbAngMom[0,2])
+        plt.xlabel('time (seconds)')
+        plt.ylabel('Relative Difference')
+        plt.title('Combined Orbital Angular Momentum')
+
+        plt.figure()
+        plt.clf()
+        plt.plot(conservationTimeData, (combinedOrbEnergy - combinedOrbEnergy[0])/combinedOrbEnergy[0])
+        plt.xlabel('time (seconds)')
+        plt.ylabel('Relative Difference')
+        plt.title('Combined Orbital Energy')
+
+    if show_plots:
+        plt.show()
+    plt.close("all")
+
+    # Testing setup
+    dirCnstAccuracy = 1E-7
+    attCnstAccuracy = 1E-4
+    accuracy = 1E-12
+    if testCase == 'Gravity':
+        np.testing.assert_allclose(psi_B1, 0, atol=dirCnstAccuracy) # gravity scenario incites larger constraint violation, but still bounded
```

Comment:
You can leave error messages when using ``assert_allclose`` so that it's obvious which part of the test failed.

Reply:
This makes sense, I've added error messages.